test/Gen.hs

```diff
@@ -0,0 +1,176 @@
+import Data.Functor.Classes (showsUnaryWith)
+import Data.Proxy
+import GHC.Stack
+import GHC.TypeLits
+import Hedgehog
+import Hedgehog.Function hiding (R, S)
+import Hedgehog.Gen
+import Hedgehog.Range
+
+-- | A generator for computations, given a higher-order generator for effectful operations, & a generator for results.
+m
+  :: forall m a
+  .  (Applicative m, Show a)
+  => (forall a . Show a => (forall a . Show a => Gen a -> Gen (With (m a))) -> Gen a -> Gen (With (m a))) -- ^ A higher-order generator producing operations using any effects in @m@.
+  -> Gen a                                                                                      -- ^ A generator for results.
+  -> Gen (With (m a))                                                                           -- ^ A generator producing computations, wrapped in 'With' for convenience.
+m with = go where
+  go :: forall a . Show a => Gen a -> Gen (With (m a))
+  go a = recursive choice [(liftWith "pure" pure . showing) <$> a] [with go a]
+
+
+genT :: Gen (T a)
+genT = T <$> integral (linear 0 100)
+
+newtype T a = T { unT :: Integer }
+  deriving (Enum, Eq, Generic, Num, Ord, Real, Vary)
+
+instance Arg (T a)
+
+instance KnownSymbol s => Show (T s) where
+  showsPrec d = showsUnaryWith showsPrec (symbolVal (Proxy @s)) d . unT
+
+a :: Gen A
+a = genT
+
+type A = T "A"
+
+b :: Gen B
+b = genT
+
+type B = T "B"
+
+e :: Gen E
+e = genT
+
+type E = T "E"
+
+r :: Gen R
+r = genT
+
+type R = T "R"
+
+s :: Gen S
+s = genT
+
+type S = T "S"
+
+w :: Gen W
+w = list (linear 0 100) genT
+
+type W = [T "W"]
+
+
+infixr 5 :.
+
+data Rec as where
+  Nil :: Rec '[]
+  (:.) :: a -> Rec as -> Rec (a ': as)
+
+forall :: (Forall g f, HasCallStack) => g -> f -> Hedgehog.Property
+forall g f = withFrozenCallStack $ Hedgehog.property (forall' g f)
+
+class Forall g f | g -> f, f -> g where
+  forall' :: HasCallStack => g -> f -> PropertyT IO ()
+
+instance Forall (Rec '[]) (PropertyT IO ()) where
+  forall' Nil = id
+
+instance (Forall (Rec gs) b, Show a) => Forall (Rec (Gen a ': gs)) (a -> b) where
+  forall' (g :. gs) f = do
+    a <- Hedgehog.forAll g
+    forall' gs (f a)
+
+
+data With a = With' { _showWith :: Int -> ShowS, getWith :: a }
+  deriving (Functor)
+
+instance Eq a => Eq (With a) where
+  (==) = (==) `on` getWith
+
+instance Applicative With where
+  pure = atom "_"
+  With' sf f <*> With' sa a = With' (\ d -> showParen (d > 10) (sf 10 . showString " " . sa 11)) (f a)
+
+instance Show (With a) where
+  showsPrec d (With' s _) = s d
+
+showing :: Show a => a -> With a
+showing = With' . flip showsPrec <*> id
+
+showingFn :: (Show a, Show b) => Fn a b -> With (a -> b)
+showingFn = With' . flip showsPrec <*> apply
+
+atom :: String -> a -> With a
+atom s = With' (\ _ -> showString s)
+
+liftWith :: String -> (a -> b) -> With a -> With b
+liftWith s w a = atom s w <*> a
+
+liftWith2 :: String -> (a -> b -> c) -> With a -> With b -> With c
+liftWith2 s w a b = atom s w <*> a <*> b
```

Comment:
The `With` DSL—named for `forAllWith`, which it subsumes—enables us to build descriptions compositionally, just like `Gen` allows us to build shrinks compositionally. (I looked into `MonadGen` to see if I could generate these by construction there; I cannot.)

- `atom` describes nullary operators.
- `liftWith` & `liftWith2` describe _and apply_ unary & binary operators, respectively.
- `showing` describes primitive values using a `Show` instance.
- `showingFn` describes functions generated using `hedgehog-fn` using `Show` instances.
- the `Applicative` instance combines the descriptions by juxtaposition, yielding descriptions of Haskell function application with appropriate precedence. At present, prefix notation is used for all applications, including infix operators (which should be written with parentheses).

As a result, if we generate a computation `local f m`, we’ll see `local f m` in the output—with one caveat that I’m aware of: the `Show` instance `hedgehog-fn` provides for `Fn` will show output such as `A 0 -> empty _ -> pure (B 1)` for a function which operates as `\case { A 0 -> empty ; _ -> pure (B 1) }`, and to the best of my knowledge it will not parenthesize it appropriately. I may submit a PR to resolve this at some point, or simply wrap `Fn` for our own purposes.

Without this, the vast majority of generated computations can’t be shown as anything more interesting than `_`, which makes debugging and resolving test failures _incredibly_ difficult. This difficulty will be significantly compounded by the extension of `m`, above, to generate monadically-sequenced chains of operations.